cmake/modules/AddSwift.cmake

```diff
@@ -2326,44 +2326,47 @@ macro(add_swift_lib_subdirectory name)
 endmacro()
 
 function(add_swift_host_tool executable)
-  set(ADDSWIFTHOSTTOOL_multiple_parameter_options
-        SWIFT_COMPONENT
+  set(options)
+  set(single_parameter_options)
+  set(multiple_parameter_options
         COMPILE_FLAGS
         DEPENDS
+        SWIFT_COMPONENT
         SWIFT_MODULE_DEPENDS)
 
-  cmake_parse_arguments(
-      ADDSWIFTHOSTTOOL # prefix
-      "" # options
-      "" # single-value args
-      "${ADDSWIFTHOSTTOOL_multiple_parameter_options}" # multi-value args
-      ${ARGN})
-
-  # Configure variables for this subdirectory.
-  set(MODULE_VARIANT_SUFFIX "-swiftmodule${SWIFT_HOST_VARIANT_SUFFIX}")
+  cmake_parse_arguments(ASHT
+                        "${options}"
+                        "${single_parameter_options}"
+                        "${multiple_parameter_options}"
+                        ${ARGN})
 
-  foreach(mod ${ADDSWIFTHOSTTOOL_SWIFT_MODULE_DEPENDS})
-    list(APPEND ADDSWIFTHOSTTOOL_DEPENDS "swift${mod}${MODULE_VARIANT_SUFFIX}")
-    list(APPEND ADDSWIFTHOSTTOOL_DEPENDS "swift${mod}${SWIFT_HOST_VARIANT_SUFFIX}")
+  foreach(mod ${ASHT_SWIFT_MODULE_DEPENDS})
+    list(APPEND ASHT_DEPENDS
+         "swift${mod}-swiftmodule${SWIFT_HOST_VARIANT_SUFFIX}"
+         "swift${mod}${SWIFT_HOST_VARIANT_SUFFIX}")
   endforeach()
 
   # Create the executable rule.
-  add_swift_executable(
-    ${executable} 
-    ${ADDSWIFTHOSTTOOL_UNPARSED_ARGUMENTS}
-    DEPENDS ${ADDSWIFTHOSTTOOL_DEPENDS}
-    COMPILE_FLAGS ${ADDSWIFTHOSTTOOL_COMPILE_FLAGS}
-  )
+  add_swift_executable(${executable}
+                       ${ASHT_UNPARSED_ARGUMENTS}
+                       DEPENDS ${ASHT_DEPENDS}
+                       COMPILE_FLAGS ${ASHT_COMPILE_FLAGS})
+  llvm_externalize_debuginfo(${executable})
+  llvm_codesign(${executable})
 
   # And then create the install rule if we are asked to.
-  if (ADDSWIFTHOSTTOOL_SWIFT_COMPONENT)
-    swift_install_in_component(${ADDSWIFTHOSTTOOL_SWIFT_COMPONENT}
-      TARGETS ${executable}
-      RUNTIME DESTINATION bin)
+  if(ASHT_SWIFT_COMPONENT)
+    swift_install_in_component(${ASHT_SWIFT_COMPONENT}
+                               TARGETS ${executable}
+                               RUNTIME DESTINATION bin
+                               COMPONENT ${executable})
+    if(NOT LLVM_ENABLE_IDE)
+      add_llvm_install_targets(install-${executable}
```

Comment:
If I passed the executable as an LLVM component to a unified build, would I get two install commands?

Also, why shouldn't LLVM_ENABLE_IDE also stop swift from installing in component? I have never seen that used in Swift. It would be useful to add documentation about or a link about what is happening here.

Reply:
The `install-` targets are specifically to wire up the `install-distribution` target.  The executable can be passed to the LLVM components list to be installed, and you do not get two install rules.

The `LLVM_ENABLE_IDE` doesn't stop that because the way that swift components work is entirely different.  The swift components actually strip out rules entirely rather than LLVM's behaviour which creates the rules but does not actually put the content into the image.

The LLVM way, IMO, is much nicer and uniform across builds and should be adopted through out swift, but I wanted to get the unified build working first (particularly as the monorepo transition looms ahead).

Reply:
For some added context, the documentation string from `LLVM_ENABLE_IDE` is `Disable optional build system features that cause problems for IDE generators`. The whole point is to improve the generation of IDE projects (mostly Visual Studio and Xcode). Both Visual Studio and Xcode have numerous UI elements and behaviors that struggle with large number of targets (think drop downs with every target in the build system). To make those projects more usable `LLVM_ENABLE_IDE=On` disables a bunch of build system features that don't contribute to correctness, but do improve command-line workflows.

Reply:
@llvm-beanz that suggests to me that swift should do the same thing and disable install rules.

Reply:
@gottesmm yeah, that would certainly be better and reduce a significant amount of complexity in the build :-).

Reply:
Yeah, I think that is the right direction forward and future work will migrate us towards the LLVM components.